refactor: move theorems about lists from mathlib

chabulhwi opened the pull request:
* `List.isEmpty_iff_eq_nil` and `List.modifyHead_modifyHead` are from `Mathlib.Data.List.Basic`.
* `List.cons_prefix_iff` is from `Mathlib.Data.List.Infix`.

chabulhwi commented:
awaiting-review

semorrison commented:
Is there a Mathlib adaptation PR ready? A lot of things are in flight at the moment so I'd like to be confident that after merging I can get everything working quickly. See e.g. what Yury did at https://github.com/leanprover/std4/pull/758.

chabulhwi commented:
> Is there a Mathlib adaptation PR ready? A lot of things are in flight at the moment so I'd like to be confident that after merging I can get everything working quickly. See e.g. what Yury did at #758.

@semorrison I opened a Mathlib PR that removes the theorems I brought to Std.

chabulhwi commented:
awaiting-review

chabulhwi commented:
awaiting-review

chabulhwi commented:
I resolved merge conflicts.

semorrison commented on Std/Data/List/Lemmas.lean:
```diff
@@ -138,6 +139,11 @@ theorem drop_left' {l₁ l₂ : List α} {n} (h : length l₁ = n) : drop n (l
 @[simp] theorem isEmpty_nil : ([] : List α).isEmpty = true := rfl
 @[simp] theorem isEmpty_cons : (x :: xs : List α).isEmpty = false := rfl
 
+-- Mathlib porting note: this does not work as desired
+-- attribute [simp] List.isEmpty
+
```
```suggestion
```

semorrison commented on Std/Data/List/Lemmas.lean:
```diff
@@ -1156,6 +1162,17 @@ theorem drop_eq_nil_of_eq_nil : ∀ {as : List α} {i}, as = [] → as.drop i =
 theorem ne_nil_of_drop_ne_nil {as : List α} {i : Nat} (h: as.drop i ≠ []) : as ≠ [] :=
   mt drop_eq_nil_of_eq_nil h
 
+/-! ### modify head -/
```
```suggestion
/-! ### modifyHead -/
```

semorrison commented on Std/Data/List/Lemmas.lean:
```diff
@@ -1156,6 +1162,17 @@ theorem drop_eq_nil_of_eq_nil : ∀ {as : List α} {i}, as = [] → as.drop i =
 
+/-! ### modify head -/
+
+-- Mathlib porting note: List.modifyHead has @[simp], and Lean 4 treats this as
+-- an invitation to unfold modifyHead in any context,
+-- not just use the equational lemmas.
+
```
```suggestion
```

semorrison commented on Std/Data/List/Lemmas.lean:
```diff
@@ -1156,6 +1162,17 @@ theorem drop_eq_nil_of_eq_nil : ∀ {as : List α} {i}, as = [] → as.drop i =
+-- an invitation to unfold modifyHead in any context,
+-- not just use the equational lemmas.
+
+-- @[simp]
```
```suggestion
```

semorrison commented on Std/Data/List/Lemmas.lean:
```diff
@@ -1156,6 +1162,17 @@ theorem drop_eq_nil_of_eq_nil : ∀ {as : List α} {i}, as = [] → as.drop i =
+-- not just use the equational lemmas.
+
+-- @[simp]
+@[simp 1100, nolint simpNF]
```
Why is this `nolint` needed? Could you add a comment explaining?

semorrison replied:
Why is the priority high? Does something else trigger first? Perhaps check whether things work removing it, and otherwise add a comment explaining?

semorrison commented on Std/Data/List/Lemmas.lean:
```diff
@@ -2352,6 +2369,15 @@ theorem IsInfix.filter (p : α → Bool) ⦃l₁ l₂ : List α⦄ (h : l₁ <:+
   obtain ⟨xs, ys, rfl⟩ := h
   rw [filter_append, filter_append]; apply infix_append _
 
+theorem cons_prefix_iff : a :: l₁ <+: b :: l₂ ↔ a = b ∧ l₁ <+: l₂ := by
```
Presumably this was already named like this in Mathlib, but I'm a little surprised this isn't `cons_prefix_cons`.

semorrison replied:
Let's rename, adding a `@[deprecated] alias`.

chabulhwi commented:
awaiting-review

eric-wieser commented:
The PR description should be a coherent summary, not a list of commit messages. If the commit messages are largely separate, then this should be separate PRs; for instance, perhaps the porting note cleanup and renames can happen in mathlib before you move things.

chabulhwi commented:
@eric-wieser It's not hard to make a summary of my commits. I'll do it in a day.

chabulhwi commented:
@eric-wieser Do the commits in a Batteries PR eventually squashed?

chabulhwi commented:
depends-on-another-PR

chabulhwi commented:
> The PR description should be a coherent summary, not a list of commit messages. If the commit messages are largely separate, then this should be separate PRs; for instance, perhaps the porting note cleanup and renames can happen in mathlib before you move things.

I split a commit into a separate PR and squashed the rest of the commits.

fgdorais commented on Batteries/Data/List/Lemmas.lean:
```diff
@@ -1217,6 +1226,17 @@ theorem IsInfix.filter (p : α → Bool) ⦃l₁ l₂ : List α⦄ (h : l₁ <:+
+  · rintro ⟨rfl, h⟩
+    rwa [prefix_cons_inj]
+
+@[deprecated] alias cons_prefix_iff := cons_prefix_cons
```
Since `cons_prefix_iff` never existed in Batteries, it doesn't make sense to deprecate here. The deprecation should be in Mathlib.

chabulhwi replied:
I moved the alias to the corresponding Mathlib PR.
